File: solution/python3/703.kth-largest-element-in-a-stream.py
#
# @lc app=leetcode id=703 lang=python3
#
# [703] Kth Largest Element in a Stream
#

# @lc code=start
class KthLargest:
    k = -1
    stream = -1

    def __init__(self, k: int, nums: List[int]):
        tmp = sorted(nums)
        self.stream = tmp[-k:]
        self.k = k

    def add(self, val: int) -> int:
        tmp = self.stream[:]
        tmp.append(val)
        self.stream = sorted(tmp)[-self.k:]
        return self.stream[0]

# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)
# @lc code=end

File: solution/java8/703.kth-largest-element-in-a-stream.java
/*
 * @lc app=leetcode id=703 lang=java
 *
 * [703] Kth Largest Element in a Stream
 */

// @lc code=start
import java.util.*;

class KthLargest {
    private Queue<Integer> queue;
    private int k;

    public KthLargest(int k, int[] nums) {
        this.k = k;
        this.queue = new PriorityQueue<Integer>();
        for (int i: nums) {
            this.add(i);
        }
    }

    public int add(int val) {
        this.queue.add(new Integer(val));
        if (this.queue.size() > this.k) {
            this.queue.poll();
        }
        System.out.println(this.queue);
        return this.queue.peek();
    }
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest obj = new KthLargest(k, nums);
 * int param_1 = obj.add(val);
 */
// @lc code=end

